mlflow/store/artifact/azure_data_lake_artifact_repo.py

```diff
@@ -0,0 +1,133 @@
+
+    match = re.match(r"([^@]+)@([^.]+)\.dfs\.core\.windows\.net", parsed.netloc)
+
+    if match is None:
+        raise MlflowException(
+            "ABFSS URI must be of the form abfss://<filesystem>@<account>.dfs.core.windows.net"
+        )
+    filesystem = match.group(1)
+    account_name = match.group(2)
+    path = parsed.path
+    if path.startswith("/"):
+        path = path[1:]
+    return filesystem, account_name, path
+
+
+def _get_data_lake_client(account_url, credential):
+    from azure.storage.filedatalake import DataLakeServiceClient
+
+    return DataLakeServiceClient(account_url, credential)
+
+
+class AzureDataLakeArtifactRepository(ArtifactRepository):
+    """
+    Stores artifacts on Azure Data Lake Storage Gen2.
+
+    This repository is used with URIs of the form
+    ``abfs[s]://file_system@account_name.dfs.core.windows.net/<path>/<path>``.
+
+    :param credential: Azure credential (see options in https://learn.microsoft.com/en-us/python/api/azure-core/azure.core.credentials?view=azure-python)
+                       to use to authenticate to storage
+    """
+
+    def __init__(self, artifact_uri, credential):
+        super().__init__(artifact_uri)
+        _DEFAULT_TIMEOUT = 600  # 10 minutes
+        self.write_timeout = MLFLOW_ARTIFACT_UPLOAD_DOWNLOAD_TIMEOUT.get() or _DEFAULT_TIMEOUT
+
+        (filesystem, account_name, path) = _parse_abfss_uri(artifact_uri)
+
+        account_url = f"https://{account_name}.dfs.core.windows.net"
+        data_lake_client = _get_data_lake_client(account_url=account_url, credential=credential)
+        self.fs_client = data_lake_client.get_file_system_client(filesystem)
+        self.base_data_lake_directory = path
+
+    def log_artifact(self, local_file, artifact_path=None):
+        raise NotImplementedError(
+            "This artifact repository does not support logging single artifacts"
+        )
+
+    def log_artifacts(self, local_dir, artifact_path=None):
+        dest_path = self.base_data_lake_directory
+        if artifact_path:
+            dest_path = posixpath.join(dest_path, artifact_path)
+        dir_client = self.fs_client.get_directory_client(dest_path)
+        local_dir = os.path.abspath(local_dir)
+        for root, _, filenames in os.walk(local_dir):
+            rel_path = os.path.relpath(root, local_dir)
+            for f in filenames:
+                # TODO: can base directory client get file at path/to/directory? or do we need
+                # a new directory client per local `root` directory that we walk in os.walk?
+                file_client = dir_client.get_file_client(posixpath.join(rel_path, f))
+                local_file_path = os.path.join(root, f)
+                if os.path.getsize(local_file_path) == 0:
+                    file_client.create_file()
+                else:
+                    with open(local_file_path, "rb") as file:
+                        file_client.upload_data(data=file, overwrite=True)
```

Comment:
Can we always run `upload_data` even when `local_file_path` is empty?

Reply:
Unfortunately not, seems no file is created in ADLS when the file being uploaded is empty